### ISSUE TITLE
Persisting data to host file system on OSX fails on startup

### ISSUE THREAD

**Issue by oobles** [docker-library/cassandra#24]
I was attempting to run Cassandra on OSX under Virtualbox and store data directly to the OSX.  docker-machine already maps /Users in the virtual machine, so I thought it would be easy to start Cassandra with the following:

docker run --name cassandra -v /Users/username/cassandra:/var/lib/cassandra/data cassandra:latest

However, on startup Cassandra reports the following and exits.  It manages to create some directories and files on the host directory, so it's not simply a permissions issue.  Is there a work around?

INFO  10:50:57 Initializing system.available_ranges
INFO  10:50:58 Enqueuing flush of local: 653 (0%) on-heap, 0 (0%) off-heap
INFO  10:50:58 Writing Memtable-local@1904409473(0.107KiB serialized bytes, 3 ops, 0%/0% of on/off-heap limit)
INFO  10:50:58 Completed flushing /var/lib/cassandra/data/system/local-7ad54392bcdd35a684174e047860b377/tmp-la-5-big-Data.db (0.000KiB) for commitlog position ReplayPosition(segmentId=1441104657062, position=283)
WARN  10:50:58 fsync(104) failed, errno (22).
WARN  10:50:58 fsync(102) failed, errno (22).
WARN  10:50:58 fsync(103) failed, errno (22).
INFO  10:50:58 Compacting (541defc0-5097-11e5-b5f0-091830ac5256) [/var/lib/cassandra/data/system/local-7ad54392bcdd35a684174e047860b377/la-5-big-Data.db:level=0, /var/lib/cassandra/data/system/local-7ad54392bcdd35a684174e047860b377/la-2-big-Data.db:level=0, /var/lib/cassandra/data/system/local-7ad54392bcdd35a684174e047860b377/la-1-big-Data.db:level=0, /var/lib/cassandra/data/system/local-7ad54392bcdd35a684174e047860b377/la-3-big-Data.db:level=0, /var/lib/cassandra/data/system/local-7ad54392bcdd35a684174e047860b377/la-4-big-Data.db:level=0, ]
INFO  10:50:58 Initializing system_distributed.parent_repair_history
INFO  10:50:58 Initializing system_distributed.repair_history
INFO  10:50:58 Initializing system_traces.sessions
INFO  10:50:58 Initializing system_traces.events
WARN  10:50:58 fsync(105) failed, errno (22).
WARN  10:50:58 fsync(107) failed, errno (22).
INFO  10:50:58 completed pre-loading (8 keys) key cache.
INFO  10:50:58 No commitlog files found; skipping replay
ERROR 10:50:58 Exception in thread Thread[CompactionExecutor:1,1,main]
org.apache.cassandra.io.FSWriteError: java.nio.file.FileSystemException: /var/lib/cassandra/data/system/local-7ad54392bcdd35a684174e047860b377/tmplink-la-6-big-Index.db -> /var/lib/cassandra/data/system/local-7ad54392bcdd35a684174e047860b377/tmp-la-6-big-Index.db: Operation not permitted
	at org.apache.cassandra.io.util.FileUtils.createHardLink(FileUtils.java:93) ~[apache-cassandra-2.2.0.jar:2.2.0]
	at org.apache.cassandra.io.sstable.format.big.BigTableWriter.makeTmpLinks(BigTableWriter.java:295) ~[apache-cassandra-2.2.0.jar:2.2.0]
	at org.apache.cassandra.io.sstable.format.big.BigTableWriter.openFinalEarly(BigTableWriter.java:332) ~[apache-cassandra-2.2.0.jar:2.2.0]
	at org.apache.cassandra.io.sstable.SSTableRewriter.switchWriter(SSTableRewriter.java:298) ~[apache-cassandra-2.2.0.jar:2.2.0]
	at org.apache.cassandra.io.sstable.SSTableRewriter.doPrepare(SSTableRewriter.java:346) ~[apache-cassandra-2.2.0.jar:2.2.0]
	at org.apache.cassandra.utils.concurrent.Transactional$AbstractTransactional.prepareToCommit(Transactional.java:169) ~[apache-cassandra-2.2.0.jar:2.2.0]
	at org.apache.cassandra.db.compaction.writers.CompactionAwareWriter.doPrepare(CompactionAwareWriter.java:79) ~[apache-cassandra-2.2.0.jar:2.2.0]
	at org.apache.cassandra.utils.concurrent.Transactional$AbstractTransactional.prepareToCommit(Transactional.java:169) ~[apache-cassandra-2.2.0.jar:2.2.0]
	at org.apache.cassandra.utils.concurrent.Transactional$AbstractTransactional.finish(Transactional.java:179) ~[apache-cassandra-2.2.0.jar:2.2.0]
	at org.apache.cassandra.db.compaction.writers.CompactionAwareWriter.finish(CompactionAwareWriter.java:89) ~[apache-cassandra-2.2.0.jar:2.2.0]
	at org.apache.cassandra.db.compaction.CompactionTask.runMayThrow(CompactionTask.java:196) ~[apache-cassandra-2.2.0.jar:2.2.0]
	at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:28) ~[apache-cassandra-2.2.0.jar:2.2.0]
	at org.apache.cassandra.db.compaction.CompactionTask.executeInternal(CompactionTask.java:74) ~[apache-cassandra-2.2.0.jar:2.2.0]
	at org.apache.cassandra.db.compaction.AbstractCompactionTask.execute(AbstractCompactionTask.java:59) ~[apache-cassandra-2.2.0.jar:2.2.0]
	at org.apache.cassandra.db.compaction.CompactionManager$BackgroundCompactionCandidate.run(CompactionManager.java:236) ~[apache-cassandra-2.2.0.jar:2.2.0]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471) ~[na:1.7.0_79]
	at java.util.concurrent.FutureTask.run(FutureTask.java:262) ~[na:1.7.0_79]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) ~[na:1.7.0_79]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [na:1.7.0_79]
	at java.lang.Thread.run(Thread.java:745) [na:1.7.0_79]
Caused by: java.nio.file.FileSystemException: /var/lib/cassandra/data/system/local-7ad54392bcdd35a684174e047860b377/tmplink-la-6-big-Index.db -> /var/lib/cassandra/data/system/local-7ad54392bcdd35a684174e047860b377/tmp-la-6-big-Index.db: Operation not permitted
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:91) ~[na:1.7.0_79]
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102) ~[na:1.7.0_79]
	at sun.nio.fs.UnixFileSystemProvider.createLink(UnixFileSystemProvider.java:475) ~[na:1.7.0_79]
	at java.nio.file.Files.createLink(Files.java:1039) ~[na:1.7.0_79]
	at org.apache.cassandra.io.util.FileUtils.createHardLink(FileUtils.java:89) ~[apache-cassandra-2.2.0.jar:2.2.0]
	... 19 common frames omitted

**Comment by oobles** [comment]
It looks like this is caused by VirtualBox not allowing linking on vboxfs.  I've added a comment to the virtualbox related ticket as symbolic linking is now supported, but hard links don't seem to be.

https://www.virtualbox.org/ticket/10085#comment:41

Also attempted the same on vmware fusion (using vmhgfs) and found the same issue.  Symbolic links are supported but hard links fails.

The only useful work around to this seems to be using NFS as documented at

http://blog.nchaulet.fr/boot2docker-nfs-on-mac/

On the mac I shared a data directory:
sudo mkdir -p /data
sudo echo "/data -mapall=`whoami`:staff `docker-machine ip`" | sudo tee -a /etc/exports  
sudo nfsd restart  

Then in the docker machine (via docker-machine ssh dev):
sudo su  
echo "#! /bin/sh  
sleep 1  
sudo mkdir -p /data
sudo /usr/local/etc/init.d/nfs-client start  
sudo mount 192.168.59.3:/data /data -o rw,async,noatime,rsize=32768,wsize=32768,proto=tcp" > /var/lib/boot2docker/bootlocal.sh  
sudo chmod a+x /var/lib/boot2docker/bootlocal.sh  
exit  
exit  

After that I could run with:

docker run --name cassandra -v /data/cassandra:/var/lib/cassandra/data cassandra:latest

I'm yet to perform any functions other than startup at this stage, so may hit some other issues yet.
